File: js/q-12.js
function uniq (list) {
  return Array.from(new Set(list))
}

function detectRoot (fn) {
  let i = 1
  while (true) {
    i += 1
    const str = fn(Math.sqrt(i)).slice(0, 10)
    if (new Set(str.split('')).size === 10) {
      break
    }
  }
  return i
}

console.log(
  'with int =',
  detectRoot(root => {
    return root.toString().replace('.', '')
  })
)
console.log(
  'only fract =',
  detectRoot(root => {
    return root.toString().split('.')[1] || ''
  })
)

function withInteger () {
  let i = 1
  while (true) {
    i += 1
    const root = Math.sqrt(i)
    const str = root
      .toString()
      .replace('.', '')
      .slice(0, 10)
    if (uniq(str.split('')).length === 10) {
      break
    }
  }
  console.log(i)
}

function withFractional () {
  let i = 1
  while (true) {
    i += 1
    const root = Math.sqrt(i)
    const str = (root.toString().split('.')[1] || '').slice(0, 10)
    if (uniq(str.split('')).length === 10) {
      break
    }
  }
  console.log(i)
}
withInteger()
withFractional()

File: js/q-12.rb

def with_integer
  i = 1
  while i += 1
    str = ('%10.10f' % Math.sqrt(i)).sub('.', '')[0..9]
    break if str.split('').uniq.length == 10
  end
  i
end

def with_fractional
  i = 1
  while i += 1
    str = ('%10.10f' % Math.sqrt(i)).split('.')[1]
    break if str.split('').uniq.length == 10
  end
  i
end
p with_integer
p with_fractional
